stellargraph/core/element_data.py

```diff
@@ -331,28 +364,53 @@ def __init__(self, shared, type_starts):
 
     def _init_directed_adj_lists(self):
         # record the edge ilocs of incoming, outgoing and both-direction edges
-        in_dict = {}
-        out_dict = {}
+        if len(self.targets) == 0:
+            empty_array = np.array([], dtype=np.uint8)
+            self._edges_in_dict = FlatAdjacencyList(empty_array, empty_array)
+            self._edges_out_dict = FlatAdjacencyList(empty_array, empty_array)
+            return
+
+        # this could be less than the true number of nodes if the node with the largest iloc is isolated
+        # but this doesn't cause issues with the code below
+        number_of_nodes = max(self.targets.max(), self.sources.max()) + 1
+        dtype = np.min_scalar_type(len(self.sources))
```

Comment:
I think this `dtype` is actually the same as `self._id_index._dtype`, and that'd be more appropriate than calculating this twice? (Maybe changing `._dtype` -> `.dtype` so it's not "private"?)